### PR TITLE
monitoring: 'flatten' observables (reimplement #11985)

### PR DESCRIPTION
**Pull request by bobheadxi**
Re-introduce "flattening" of observables by wrapping queries in a `min` or `max` where appropriate (first added in https://github.com/sourcegraph/sourcegraph/pull/11985) to fix a regression introduced in https://github.com/sourcegraph/sourcegraph/issues/17599. Also add monitoring to ensure we have some visibility into rule failures, and reorganize Prometheus panels for uniformity.

image

See docstring and linked issues/PRs in https://github.com/sourcegraph/sourcegraph/issues/17599 for more context.

Main diffs to look at:

- https://github.com/sourcegraph/sourcegraph/pull/17600/files#diff-4132060621aa0b48b1ada84cd9c0fc0984c98250bfce3fe78d4f15eb5d8b39f6
- https://github.com/sourcegraph/sourcegraph/blob/monitoring/flatten-observables/doc/admin/observability/alert_solutions.md#prometheus-prometheus_rule_eval_failures
- https://github.com/sourcegraph/sourcegraph/blob/monitoring/flatten-observables/doc/admin/observability/dashboards.md#prometheus-prometheus_rule_eval_failures

Closes https://github.com/sourcegraph/sourcegraph/issues/17599

### CODE REVIEW

File: monitoring/monitoring/monitoring.go [hunk 1]
```diff
@@ -304,27 +304,49 @@ func (c *Container) renderRules() (*promRulesFile, error) {
...
+						// Comparator sets how a metric should be compared against a threshold
+						comparator string
+						// Threshold sets the value to be compared against
+						threshold float64
+					)
+
+					// Set values to build a query with
 					if a.greaterThan != nil {
-						comparator := ">="
+						aggregator = "min" // alert if the largest value is exceeds upper bound
```
Comment (pecigonzalo) [inline monitoring/monitoring/monitoring.go:323]:
```suggestion
						aggregator = "max" // alert if the largest value is exceeds upper bound
```
Text says the largest

**Comment by pecigonzalo** [comment]
I would honestly prefer to remove the `min` and `max` 🤔 I wonder what is causing this issue as I have not had to wrap things up like this in general.

In any case, I would merge to fix the issues until we figure that out.

(not approving as I believe there is a typo which I mentioned)

File: monitoring/monitoring/monitoring.go (hunk 1 above)
Comment (bobheadxi) [reply monitoring/monitoring/monitoring.go:323]: oops you're right!

**Comment by bobheadxi** [comment]
> I would honestly prefer to remove the min and max 🤔

I think they're fine to add, not not fantastic, thing we are in the end achieving the same thing: alerting on the first occurrence of a value exceeding a threshold

> I wonder what is causing this issue as I have not had to wrap things up like this in general.

This describes it well I think: https://github.com/sourcegraph/sourcegraph/issues/11571#issuecomment-654571953

**Comment by pecigonzalo** [comment]
Thanks for that link, yeah, but we should not have those collisions, or it is rather rare, so I wonder if we have the wrong query or the wrong collection rule or something else altogether. Or maybe this is just more normal than I expect it to be :D

**Comment by bobheadxi** [comment]
Our alert delivery, formatting, and entire model does not support per-series anyway, and there's a [whole monitoring pillars section](https://about.sourcegraph.com/handbook/engineering/observability/monitoring_pillars#graphs-should-have-less-than-5-cardinality) about why high-cardinality graphs are bad.

If we want individual series alerting, I think they should just be separate panels, and instead of surfacing information through the alert itself IMO we should just push to have useful information about the panel be surfaced instead (e.g. a link to it, where more context can be derived), so I think this change is inline with that

**Review by pecigonzalo — APPROVED** [review]

**Pull request merged**